src/engine.rs

```diff
@@ -965,20 +1017,112 @@ impl Reedline {
+                        // TODO properly adjusting prompt_origin on resizing
+                        // take into account multiline
+                        prompt_origin.1 = position()?.1.saturating_sub(1);
+                        prompt_offset = self.full_repaint(prompt_origin, width)?;
+                        continue;
                     }
                 }
-                if self.history_search.is_some() {
-                    self.history_search_paint()?;
-                } else if self.need_full_repaint {
-                    self.full_repaint(prompt_origin, terminal_size.0)?;
+                if self.need_full_repaint {
+                    prompt_offset = self.full_repaint(prompt_origin, terminal_size.0)?;
                     self.need_full_repaint = false;
                 } else {
                     self.buffer_paint(prompt_offset)?;
                 }
             } else {
-                self.full_repaint(prompt_origin, terminal_size.0)?;
+                prompt_offset = self.full_repaint(prompt_origin, terminal_size.0)?;
+            }
+        }
+    }
+}
+
+/// Return the index up until which the String is displayable in the given width
+/// If the string is displayable in its entirety -> return `str::len()`
+/// ` rust
+/// let str_full = "display me";
+/// let str_short = &str_full[..str_index_at_width(&str_full, 8)];
+/// `
+fn str_index_at_width(strr: &str, width: usize) -> usize {
+    use unicode_width::UnicodeWidthChar;
+    if strr.len() < width {
+        return strr.len();
+    }
+
+    let char_indices = strr.char_indices();
+    let mut state = 0;
+    for (idx, chr) in char_indices {
+        state += chr.width().unwrap_or(0);
+        if state > width {
+            return idx;
+        } else {
+            continue;
+        }
+    }
+    strr.len()
+}
+
+// Repaint logic for the interactive history search
+// Overwrites the prompt indicator and highlights the search string
+// separately from the result buffer.
+fn build_list<T: AsRef<str>>(
+    items: impl Iterator<Item = T>,
+    search_string: &str,
+    list_index: usize,
+) -> Result<(String, usize, Option<usize>)> {
+    fn push_entry(list: &mut String, entry: &str, remaining_cols: usize) {
+        if entry.len() < remaining_cols as usize {
+            //ascii fast-path
+            list.push_str(entry);
+        } else {
+            let entry_trimmed = &entry[..str_index_at_width(&entry, remaining_cols)];
+            list.push_str(entry_trimmed);
+            if entry_trimmed.len() < entry.len() {
+                list.push('.');
+                list.push('.');
             }
         }
     }
+    let mut list = String::with_capacity(8 * 32);
+
+    let mut search_results = items
+        .enumerate()
+        .skip(list_index.saturating_sub(7))
+        .filter(|(_, entry)| entry.as_ref().starts_with(&*search_string))
+        .enumerate()
+        .take(8)
+        .peekable();
```

Comment:
Maybe I will call it run_search_and_build_list, because the filtering needs to be done repeatedly (on changes to the search_string) and the implementation depends on the different indices that are created through enumerate. 
If I factor out even more functionality, only the pushing of the different entries to the list would be left.